corehq/apps/domain/auth.py

```diff
@@ -81,22 +78,21 @@ def _decode(string):
             return string.decode('latin1')
 
     username = password = None
-    if 'HTTP_AUTHORIZATION' in request.META:
-        auth = request.META['HTTP_AUTHORIZATION'].split()
-        if auth[0].lower() == DIGEST:
-            try:
-                digest = parse_digest_credentials(request.META['HTTP_AUTHORIZATION'])
-                username = digest.username
-            except UnicodeDecodeError:
-                pass
-        elif auth[0].lower() == BASIC:
-            username, password = base64.b64decode(auth[1]).split(b':', 1)
-            # decode password submitted from mobile app login
-            password = decode_password(password)
-            username, password = _decode(username), _decode(password)
-    else:
-        # Return the username for API requests that pass it as a GET parameter
-        username = request.GET['username']
+    auth_type = determine_authtype_from_header(request, default='NONE')
+    if auth_type == DIGEST:
+        try:
+            digest = parse_digest_credentials(request.META['HTTP_AUTHORIZATION'])
+            username = digest.username
+        except UnicodeDecodeError:
+            pass
+    elif auth_type == BASIC:
+        _, credentials = request.META['HTTP_AUTHORIZATION'].split()
+        username, password = base64.b64decode(credentials).split(b':', 1)
+        # decode password submitted from mobile app login
+        password = decode_password(password)
+        username, password = _decode(username), _decode(password)
+    elif auth_type == API_KEY:
+        username, _ = ApiKeyAuthentication().extract_credentials(request)
```

Comment:
Nice, I'm starting to like the structure of these functions.  Feels like we're trending toward a place where various auth methods are cleanly abstracted away.

Reply:
Hmm, with the exception of standard session auth, that is.  Do you know how that'd fit in here?